File: packages/generator-go/detail_test.go
package generator

import "testing"

func TestBuildDetailUsesSharedDisplayEngine(t *testing.T) {
	vm, err := BuildDetail(NewObject("fields", NewObject(
		"name", NewObject("field", ".name", "label", "Name"),
		"state", NewObject("field", ".state", "format", NewObject("type", "badge", "map", NewObject("active", "success"))),
	)), NewObject("name", "Ada", "state", "active"), DetailOptions{})
	if err != nil {
		t.Fatal(err)
	}
	fields := objectList(read(vm, "fields"))
	if len(fields) != 2 || stringAt(fields[0], "display") != "Ada" {
		t.Fatalf("unexpected detail fields: %#v", fields)
	}
	if stringAt(object(read(fields[1], "display")), "kind") != "badge" {
		t.Fatalf("expected badge display: %#v", fields[1])
	}
}

func TestRenderDetailIsReadOnly(t *testing.T) {
	html, err := RenderDetail(NewObject("fields", NewObject("name", NewObject("field", ".name", "label", "Name"))), NewObject("name", "Ada"), DetailOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if html != `<dl class="crudui-detail"><div class="crudui-detail__field"><dt class="crudui-detail__label">Name</dt><dd class="crudui-detail__value crudui-value crudui-value--text">Ada</dd></div></dl>` {
		t.Fatalf("unexpected detail HTML: %s", html)
	}
}

File: packages/generator-core/src/detail.test.ts
import { describe, expect, test } from 'vitest';
import { buildDetail } from './detail';

describe('buildDetail', () => {
  test('builds ordered read-only fields through the list display engine', () => {
    const vm = buildDetail({
      fields: {
        name: { field: 'name', label: { en: 'Name', ko: '이름' } },
        state: { field: 'state', format: { type: 'badge', map: { active: 'success' } } },
      },
    }, { name: 'Ada', state: 'active' }, { language: 'en' });

    expect(vm.fields.map((field) => field.key)).toEqual(['name', 'state']);
    expect(vm.fields[0]?.display).toBe('Ada');
    expect(vm.fields[1]?.display).toEqual({ kind: 'badge', variant: 'success', label: 'active' });
  });

  test('rejects a declaration or record that is not an object', () => {
    for (const spec of [null, 'detail', ['fields'], 1]) {
      expect(() => buildDetail(spec as never)).toThrow('Detail specification must be an object');
    }
    expect(() => buildDetail({})).toThrow('Detail specification must declare fields');
    expect(() => buildDetail({ fields: {} }, [] as never))
      .toThrow('Detail record must be an object');
  });

  test('carries the same cell members a list row carries', () => {
    const vm = buildDetail({ fields: { name: { field: 'name' } } }, { name: 'Ada' });
    expect(Object.keys(vm.fields[0] ?? {})).toEqual(['key', 'label', 'format', 'value', 'display', 'design']);
  });

  test('resolves composed detail fields', () => {
    const vm = buildDetail({ fields: { $ref: 'shared.json' } }, { name: 'Ada' }, {
      files: { 'shared.json': { properties: { name: { field: 'name' } } } },
    });
    expect(vm.fields[0]?.display).toBe('Ada');
  });
});
